iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/azure/AzurePublicNetworkAccessCheckPart.java

```diff
@@ -52,21 +49,44 @@ protected void registerResourceChecks() {
       "azurerm_postgresql_server",
       "azurerm_redis_cache",
       "azurerm_search_service",
-      "azurerm_synapse_workspace");
-    register(AzurePublicNetworkAccessCheckPart::checkPublicNetworkVariant, "azurerm_data_factory", "azurerm_purview_account");
+      "azurerm_synapse_workspace"), reportEnabledPublicIp("public_network_access_enabled"));
+    resourceConsumer(List.of("azurerm_data_factory", "azurerm_purview_account"), reportEnabledPublicIp("public_network_enabled"));
+
+    resourceConsumer("azurerm_application_gateway", reportPublicIpConfiguration("frontend_ip_configuration"));
+    resourceConsumer("azurerm_network_interface", reportPublicIpConfiguration("ip_configuration"));
+
+    resourceConsumer(List.of("azurerm_dev_test_linux_virtual_machine", "azurerm_dev_test_windows_virtual_machine"),
+      resource -> resource.attribute("disallow_public_ip_address")
+        .reportOnFalse(NETWORK_ACCESS_MESSAGE)
+        .reportAbsence(OMITTED_MESSAGE));
+
+    resourceConsumer("azurerm_dev_test_virtual_network",
+      resource -> resource.block("subnet").ifPresent(
+        subnet -> subnet.attribute("use_public_ip_address")
+          .reportUnexpectedValue("Deny", NETWORK_ACCESS_MESSAGE)
+          .reportAbsence(OMITTED_MESSAGE)));
+
+    resourceConsumer("azurerm_batch_pool",
+      resource -> resource.block("network_configuration").ifPresent(
+        configuration -> configuration.attribute("public_address_provisioning_type")
+          .reportUnexpectedValue("NoPublicIPAddresses", NETWORK_ACCESS_MESSAGE)));
+
+    resourceConsumer("azurerm_kubernetes_cluster_node_pool",
+      resource -> resource.block("default_node_pool").ifPresent(
+        pool -> pool.attribute("enable_node_public_ip").reportOnTrue(NETWORK_ACCESS_MESSAGE)));
   }
 
-  private static void checkPublicNetworkAccess(CheckContext ctx, BlockTree resource) {
-    PropertyUtils.get(resource, PUBLIC_NETWORK_ACCESS_ENABLED, AttributeTree.class)
-      .ifPresentOrElse(
-        enabled -> reportOnTrue(ctx, enabled, NETWORK_ACCESS_MESSAGE),
-        () -> reportResource(ctx, resource, String.format(OMITTED_MESSAGE, PUBLIC_NETWORK_ACCESS_ENABLED)));
+
+  private Consumer<Resource> reportEnabledPublicIp(String propertyName) {
+    return resource -> resource.attribute(propertyName)
+      .reportOnTrue(NETWORK_ACCESS_MESSAGE)
+      .reportAbsence(OMITTED_MESSAGE);
   }
 
-  private static void checkPublicNetworkVariant(CheckContext ctx, BlockTree resource) {
-    PropertyUtils.get(resource, PUBLIC_NETWORK_ENABLED, AttributeTree.class)
-      .ifPresentOrElse(
-        enabled -> reportOnTrue(ctx, enabled, NETWORK_ACCESS_MESSAGE),
-        () -> reportResource(ctx, resource, String.format(OMITTED_MESSAGE, PUBLIC_NETWORK_ENABLED)));
+  private Consumer<Resource> reportPublicIpConfiguration(String propertyName) {
+    return resource -> resource.blocks(propertyName).forEach(
+      gateway -> gateway.attribute("public_ip_address_id")
```

Comment:
I think I would rename `gateway` to something like `property` or `block` because this method is quite generic.